### PR TITLE
refactor(zetaclient): streamline process initialization

### PR DESCRIPTION
**Pull request by swift1337**
This PR streamlines the `start.go` file and introduces the `graceful` package to facilitate a graceful shutdown. It also includes various other miscellaneous improvements.

However, the diff is quite challenging to navigate. I suggest checking out the branch to see the changes in `start.go` and `pkg/graceful` (I also like Github Dev https://github.dev/zeta-chain/node/pull/3291)

- Closes https://github.com/zeta-chain/node/issues/3112
- Closes https://github.com/zeta-chain/node/issues/3108

## Release Notes

- **New Features**
	- Added a new telemetry endpoint `/systemtime` in the ZetaClient.
	- Introduced a function to resolve the database path for the chain observer.

- **Bug Fixes**
	- Fixed issues with unsupported transaction versions for Solana.
	- Improved error handling for various operations, including configuration loading and logger creation.

- **Improvements**
	- Enhanced initialization process for the ZetaClient with graceful shutdown capabilities.
	- Streamlined logging setup and error reporting across multiple components.

- **Tests**
	- Updated test suite for the graceful shutdown process, ensuring robust validation of service behaviors.

- **Documentation**
	- Updated changelog to reflect recent changes and improvements across features and fixes.

### CODE REVIEW

File: pkg/graceful/graceful.go [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+// AddStarter runs a function that starts something.
+// This is a blocking call for blocking .Start() services
+func (p *Process) AddStarter(ctx context.Context, fn func(ctx context.Context) error) {
+	go func() {
+		defer func() {
+			if r := recover(); r != nil {
+				p.logger.Error().Interface("panic", r).Msg("panic in service")
+				p.ShutdownNow()
+			}
+		}()
```
Comment (gartnera) [inline pkg/graceful/graceful.go:66]: Please make sure to log the actual panic source like https://github.com/zeta-chain/node/pull/2881

File: cmd/zetaclientd/utils.go [hunk 2]
```diff
@@ -97,9 +80,68 @@ func waitForZetacoreToCreateBlocks(ctx context.Context, zc interfaces.ZetacoreCl
...
+		logger.Info().Msgf("Failed to get block number, retry: %d/%d", retryCount, attempts)
+		return nil
+	}
+
+	return ticker.Run(ctx, interval, task)
+}
+
+// prepareZetacoreClient prepares the zetacore client for use.
+// EXITS THE PROGRAM IF THIS NODE IS NOT AN OBSERVER.
+func prepareZetacoreClient(ctx context.Context, zc *zetacore.Client, cfg *config.Config, logger zerolog.Logger) error {
```
Comment (kingpinXD) [inline cmd/zetaclientd/utils.go:92]: consider moving create and prepare functions to a separate file called zetacore_client. 
They don't seem like utility functions

File: cmd/zetaclientd/utils.go [hunk 3]
```diff
@@ -97,9 +80,68 @@ func waitForZetacoreToCreateBlocks(ctx context.Context, zc interfaces.ZetacoreCl
...
+			Str("want", network).
+			Msg("Zetacore chain id config mismatch. Forcing update from the network")
+
+		cfg.ChainID = network
+		if err = zc.UpdateChainID(cfg.ChainID); err != nil {
+			return errors.Wrap(err, "failed to update chain id")
+		}
+	}
+
+	isObserver, err := isObserverNode(ctx, zc)
```
Comment (kingpinXD) [inline cmd/zetaclientd/utils.go:116]: Wont this check new nodes from taking part in the TSS ceremony .
Conisder the situation when a new node is added , and is part of the node account list but not part of the observer set . 
They need to sign on the TSS ceremony.

I think this check is correct , but it should be done after the TSS voting is executed

File: cmd/zetaclientd/utils.go [hunk 4]
```diff
@@ -97,9 +80,68 @@ func waitForZetacoreToCreateBlocks(ctx context.Context, zc interfaces.ZetacoreCl
...
+	switch {
+	case err != nil:
+		return errors.Wrap(err, "failed to check if this node is an observer")
+	case !isObserver:
+		addr := zc.GetKeys().GetOperatorAddress().String()
+		logger.Info().Str("operator_address", addr).Msg("This node is not an observer. Exit 0")
+		os.Exit(0)
```
[the remark this answers is not in the document]
Comment (kingpinXD) [reply cmd/zetaclientd/utils.go:123]: I am also curious: why use exit instead of returning an error?

File: cmd/zetaclientd/utils.go [hunk 5]
```diff
@@ -97,9 +80,68 @@ func waitForZetacoreToCreateBlocks(ctx context.Context, zc interfaces.ZetacoreCl
...
+		return false, errors.Wrap(err, "unable to get observers list")
+	}
+
+	operatorAddress := zc.GetKeys().GetOperatorAddress().String()
+
+	for _, observer := range observers {
+		if observer == operatorAddress {
+			return true, nil
+		}
+	}
```
Comment (kingpinXD) [inline cmd/zetaclientd/utils.go:142]: Consider adding a function for observerSet, something like IsObserver(). 
We can put the function under x/observer/types.

File: cmd/zetaclientd/utils.go (hunk 3 above)
Comment (swift1337) [reply cmd/zetaclientd/utils.go:116]: Okay, initially it was added here
https://github.com/zeta-chain/node/pull/1031/files#diff-4f6dad4b36c4f2c7c015aaa31c4468e0fd88293d9fbf4a5f9c1be1751ea5ebe6R184

I just refactored the semantics. Will move below, thanks.

File: cmd/zetaclientd/utils.go (hunk 5 above)
Comment (swift1337) [reply cmd/zetaclientd/utils.go:142]: Same as here https://github.com/zeta-chain/node/pull/3291#discussion_r1883839054

Comment (swift1337) [reply cmd/zetaclientd/utils.go:142]: I realised that it's `QueryObserverSetResponse`, not `ObserverSet` that it used within the keeper. Will keep as is

**Review by kingpinXD — APPROVED** [review]

File: zetaclient/zetacore/client_start.go [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+import (
+	"context"
+	"fmt"
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog"
+
+	authz2 "github.com/zeta-chain/node/pkg/authz"
```
Comment (lumtis) [inline zetaclient/zetacore/client_start.go:12]:
```suggestion
	zetaauthz "github.com/zeta-chain/node/pkg/authz"
```

**Review by lumtis — APPROVED** [review]

**Pull request merged**